### PR TITLE
net: ethernet: Support HW MAC address filtering

### PR DESCRIPTION
**Pull request by pdgendt**
An existing ethernet configuration option can be utilized for MAC address filtering, this PR demonstrates this.

### CODE REVIEW

**Comment by jukkar** [comment]
Makes sense, +1 from me.

**Comment by pdgendt** [comment]
Adding DNM for now as I want some feedback for the drivers I can't test.

File: drivers/ethernet/eth_stm32_hal.c [hunk 1]
```diff
@@ -1575,6 +1431,9 @@ static enum ethernet_hw_caps eth_stm32_hal_get_capabilities(const struct device
 #endif
 #if defined(CONFIG_NET_DSA)
 		| ETHERNET_DSA_MASTER_PORT
+#endif
+#if defined(CONFIG_ETH_STM32_MULTICAST_FILTER)
+		| ETHERNET_FW_FILTER
```
Comment (jukkar) [inline drivers/ethernet/eth_stm32_hal.c:1436]: s/FW_FILTER/HW_FILTERING/

Comment (pdgendt) [reply drivers/ethernet/eth_stm32_hal.c:1436]: A `build_all` test would be nice to have.

Comment (jukkar) [reply drivers/ethernet/eth_stm32_hal.c]: There is the `tests/net/all` where suitable options could be added.

Comment (pdgendt) [reply drivers/ethernet/eth_stm32_hal.c]: @jukkar I tried to implement a `build_all` case for the stm32 driver, but it doesn't seem to take work as expected.

Could you take a look at the last commit to see what I'm missing, build output:

```log
warning: ETH_DRIVER (defined at drivers/ethernet/Kconfig:6) was assigned the value 'y' but got the
value 'n'. Check these unsatisfied dependencies: (!NET_TEST) (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_ETH_DRIVER and/or look up ETH_DRIVER in the
menuconfig/guiconfig interface. The Application Development Primer, Setting Configuration Values,
and Kconfig - Tips and Best Practices sections of the manual might be helpful too.


warning: ETH_STM32_MULTICAST_FILTER (defined at drivers/ethernet/Kconfig.stm32_hal:204) was assigned
the value 'y' but got the value 'n'. Check these unsatisfied dependencies: ETH_STM32_HAL (=n),
ETH_DRIVER (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_ETH_STM32_MULTICAST_FILTER and/or look up
ETH_STM32_MULTICAST_FILTER in the menuconfig/guiconfig interface. The Application Development
Primer, Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual
might be helpful too.


warning: PHY_INIT_PRIORITY (defined at drivers/ethernet/phy/Kconfig:16) was assigned the value '50'
but got the value ''. Check these unsatisfied dependencies: ETH_DRIVER (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_PHY_INIT_PRIORITY and/or look up
PHY_INIT_PRIORITY in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.

```

**Comment by pdgendt** [comment]
I've moved the test fixes to a separate PR here: #69784

**Review by jukkar — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Comment by Dat-NguyenDuy** [comment]
Hi @pdgendt, i would like to test on nxp s32 boards, could you tell me how you did test this?

**Comment by pdgendt** [comment]
> Hi @pdgendt, i would like to test on nxp s32 boards, could you tell me how you did test this?

Hey @Dat-NguyenDuy, thanks for your time.

I think the easiest way is to build an IPv6 sample, like `samples/net/sockets/coap_server` and verify that the driver is reachable on registered multicast addresses.

```shell
$ ping -6 -I eth0 ff02::1  # ping all nodes
64 bytes from fe80::3ab1:9eff:fe4c:c99: seq=0 ttl=64 time=5.870 ms
$ ping -6 -I eth0 ff02::fd # ping CoAP routers
64 bytes from fe80::3ab1:9eff:fe4c:c99: seq=2 ttl=64 time=4.197 ms
```

**Review by Dat-NguyenDuy — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**